lib/THC/THCTensorIndex.cu

```diff
@@ -94,6 +95,124 @@ __global__ void indexCopyLargeIndex(TensorInfo<T, IndexType> dst,
+    unsigned int * address_as_ui =
+        (unsigned int *) (address - ((size_t)address & 3));
+    unsigned int old = *address_as_ui, sum, newval, assumed;
+    unsigned int selectors[] = {0x3214, 0x3240, 0x3410, 0x4210};
+    unsigned int sel = selectors[(size_t)address & 3];
+
+    do {
+      assumed = old;
+      sum = val +  (T)__byte_perm(old, 0, (size_t)address & 3 | 0x4440);
+      newval = __byte_perm(old, sum, sel);
+      old = atomicCAS(address_as_ui, assumed, newval);
+    } while (assumed != old);
+  }
+};
+
+template<typename T>
+struct AtomicAddIntegerImpl<T, 2> {
+  __device__ void operator()(T *address, T val) {
+    unsigned int * address_as_ui =
+        (unsigned int *) ((char *)address - ((size_t)address & 2));
+    unsigned int old = *address_as_ui, sum, newval, assumed;
+
+    do {
+      assumed = old;
+      sum = val +  (T)__byte_perm(old, 0, ((size_t)address & 2) ? 0x4432 : 0x4410);
+      newval = __byte_perm(old, sum, ((size_t)address & 2) ? 0x5410 : 0x3254);
+      old = atomicCAS(address_as_ui, assumed, newval);
+    } while (assumed != old);
+  }
+};
+
+template<typename T>
+struct AtomicAddIntegerImpl<T, 4> {
+  __device__ void operator()(T *address, T val) {
+    unsigned int * address_as_ui = (unsigned int *) (address);
+    unsigned int old = *address_as_ui, newval, assumed;
+
+    do {
+      assumed = old;
+      newval = val +  (T)old;
+      old = atomicCAS(address_as_ui, assumed, newval);
+    } while (assumed != old);
+  }
+};
+
+template<typename T>
+struct AtomicAddIntegerImpl<T, 8> {
+  __device__ void operator()(T *address, T val) {
+    unsigned long long * address_as_ui = (unsigned long long *) (address);
+    unsigned long long old = *address_as_ui, newval, assumed;
+
+    do {
+      assumed = old;
+      newval = val +  (T)old;
+      old = atomicCAS(address_as_ui, assumed, newval);
+    } while (assumed != old);
+  }
+};
+
+__device__ void atomicAdd(unsigned char *address, unsigned char val) {
+  AtomicAddIntegerImpl<unsigned char, sizeof(unsigned char)>()(address, val);
+}
+
+__device__ void atomicAdd(char *address, char val) {
+  AtomicAddIntegerImpl<char, sizeof(char)>()(address, val);
+}
+
+__device__ void atomicAdd(short *address, short val) {
+  AtomicAddIntegerImpl<short, sizeof(short)>()(address, val);
+}
+
+__device__ void atomicAdd(long *address, long val) {
+  AtomicAddIntegerImpl<long, sizeof(long)>()(address, val);
+}
+
+#ifdef CUDA_HALF_INSTRUCTIONS
+#define HALF_ADD(a, b) __hadd(a, b)
+#else
+#define HALF_ADD(a, b) __float2half(__half2float(a) + __half2float(b))
+#endif
+
+__device__ void atomicAdd(half *address, half val) {
+  unsigned int * address_as_ui =
+      (unsigned int *) ((char *)address - ((size_t)address & 2));
+  unsigned int old = *address_as_ui, assumed;
+  half hsum;
+
+  do {
+    // byte_perm doesn't work with halfs, let's manually memcpy
+    assumed = old;
+    half *hp = (half *)((char *)(&old) + ((size_t)address & 2));
```

Comment:
I don't have high hopes here that the compiler is able to statically determine that lines 194-196 is equivalent to bit manipulation. Taking the address of a value in a register and performing this manipulation is liable to move the value from being in a GPU register and will force it to local memory, which is slow.

Since there are only 2 possible offset states (first 16-bit word or second 16-bit word), I would do something like this:

```
half hsum;
// I think GPUs are little endian?
hsum.x = address & 2 == 0 ? (old & 0xffff) : (old >> 16);
hsum = THCNumerics<half>::add(hsum, val);
// write the 16-bit word back into `old` and then do the CAS stuff
```

predication on the GPU should make the `? :` pretty efficient, there are only two states here. You could avoid the predication as well by determining the shift required by inspecting `address`.